perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ExtSourcesManagerBlImpl.java

```diff
@@ -62,67 +74,111 @@ public ExtSourcesManagerBlImpl(ExtSourcesManagerImplApi extSourcesManagerImpl) {
 	@Override
 	public void initialize(PerunSession sess) {
 		if (!this.initialized.compareAndSet(false, true)) return;
+		ExtSourceGoogle.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceVOOT.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceUnity.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceCSV.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceISMU.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceLdap.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceSql.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourcePerun.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceREMS.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceTCS.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceSqlComplex.setPerunBlImpl((PerunBlImpl) perunBl);
+		ExtSourceXML.setPerunBlImpl((PerunBlImpl) perunBl);
 		this.extSourcesManagerImpl.initialize(sess);
+		cacheExtSourcesInMemory(sess);
+	}
+
+	private Map<Integer,ExtSource> extSourcesByIdMap = new HashMap<>();
+	private Map<String,ExtSource> extSourcesByNameMap = new HashMap<>();
+	
+	private void cacheExtSourcesInMemory(PerunSession sess) {
+		Map<Integer,ExtSource> extSourcesByIdMap = new HashMap<>();
+		Map<String,ExtSource> extSourcesByNameMap = new HashMap<>();
+		// read all ExtSources from database
+		List<ExtSource> extSources = getExtSourcesManagerImpl().getExtSources(sess);
+		// cache them in hashmaps
+		for(ExtSource extSource : extSources) {
+			extSourcesByIdMap.put(extSource.getId(), extSource);
+			extSourcesByNameMap.put(extSource.getName(), extSource);
+		}
+		this.extSourcesByIdMap = extSourcesByIdMap;
+		this.extSourcesByNameMap = extSourcesByNameMap;
+	}
+
+	@Override
+	public void destroy() {
+		this.extSourcesManagerImpl.destroy();
 	}
 
 	@Override
 	public ExtSource createExtSource(PerunSession sess, ExtSource extSource, Map<String, String> attributes) throws InternalErrorException, ExtSourceExistsException {
 		getPerunBl().getAuditer().log(sess, new ExtSourceCreated(extSource));
-		return getExtSourcesManagerImpl().createExtSource(sess, extSource, attributes);
+		ExtSource extSource1 = getExtSourcesManagerImpl().createExtSource(sess, extSource, attributes);
+		extSourcesByIdMap.put(extSource1.getId(), extSource);
+		extSourcesByNameMap.put(extSource1.getName(), extSource);
```

Comment:
Is there any reason to use old extSource instead of returned one? If yes, than at least change name of the variable to 'returnedExtSource' or 'createdExtSource'.

Reply:
There returned class is a subclass of ExtSource based on the specified type.